configfile: "config.yaml"
IDS, = glob_wildcards("art_reads/{sample}_1.fq")

rule all:
    input:
        expand("quast/quast_spades-{sample}/report.tsv", sample = IDS),
        expand("quast/quast_skesa-{sample}/report.tsv", sample= IDS),
        expand("quast/quast_shovill_velvet-{sample}/report.tsv", sample = IDS),
        expand("quast/quast_shovill_spades-{sample}/report.tsv", sample = IDS),
        expand("assemblies/skesa/{sample}.fasta", sample = IDS),
        expand("assemblies/shovill_velvet/shovill_velvet-{sample}/velvet.fasta", sample = IDS),
        expand("assemblies/shovill_spades/shovill_spades-{sample}/spades.fasta", sample = IDS),
        expand("assemblies/spades/spades-{sample}/contigs.fasta", sample = IDS),
        "plots/NGA50.pdf",
        "plots/contigs.pdf",
        "plots/misassemblies.pdf",
        "plots/genome_fraction.pdf",
        "plots/complete_features.pdf",
        "plots/total_features.pdf",
        "ska_output/ska_spades_alignment.aln",
        "ska_output/ska_shovill_alignment.aln",
        "ska_output/ska_skesa_alignment.aln",
        "ska_output/ska_fastq_alignment.aln",
        "pirate/skesa/core_alignment/pangenome_alignment.fasta",
        "pirate/shovill/core_alignment/pangenome_alignment.fasta",
        "pirate/spades/core_alignment/pangenome_alignment.fasta",
        "roary/spades/core_gene_alignment.aln",
        "roary/shovill/core_gene_alignment.aln",
        "roary/skesa/core_gene_alignment.aln",
        "snippy-core_out/core.full.aln",
        "snippy-core_out/core.aln",
        "kSNP_shovill/core_SNPs_matrix"
        "kSNP_spades/core_SNPs_matrix"
        "kSNP_skesa/core_SNPs_matrix"

##denovo assembly with Skesa
rule skesa:
    input:
        read1 = "art_reads/{sample}_1.fq",
        read2 = "art_reads/{sample}_2.fq"
    output:
        output = "assemblies/skesa/{sample}.fasta"
    params:
        prefix = "{sample}.fasta",
        contig = config["skesa"]["min_contig_len"]
    conda:
        "envs/skesa.yaml"
    log:
        "logs/skesa/{sample}.log"
    threads: 8
    shell:
         """
         skesa --cores {threads} --min_contig {params.contig} --fastq {input.read1},{input.read2} --contigs_out {output.output} 2>&1> {log}
         """

##denovo assembly with shovill+velvet
rule shovill_velvet:
    input:
        read1 = "art_reads/{sample}_1.fq",
        read2 = "art_reads/{sample}_2.fq"
    output:
        output = "assemblies/shovill_velvet/shovill_velvet-{sample}/velvet.fasta"
    params:
        prefix = "{sample}",
        assembler = config["shovill_velvet"]["assembler"],
        len = config["shovill_velvet"]["min_contig_len"]
    conda:
        "envs/shovill.yaml"
    log:
        "logs/shovill_velvet/{sample}.log"
    threads: 8
    shell:
        """     
        shovill --force --cpus {threads} --assembler {params.assembler} --minlen {params.len} -R1 {input.read1} -R2 {input.read2} --outdir shovill_velvet-{params.prefix} 2>&1> {log}
        mv shovill_velvet-{params.prefix} assemblies/shovill_velvet
        """
##denovo assembly with shovill+spades
rule shovill_spades:
    input:
        read1 = "art_reads/{sample}_1.fq",
        read2 = "art_reads/{sample}_2.fq"
    output:
        output = "assemblies/shovill_spades/shovill_spades-{sample}/spades.fasta"
    params:
        prefix = "{sample}",
        assembler = config["shovill_spades"]["assembler"],
        len = config["shovill_spades"]["min_contig_len"]
    conda:
        "envs/shovill.yaml"
    log:
        "logs/shovill_spades/{sample}.log"
    threads: 8
    shell:
        """     
        shovill --force --cpus {threads} --assembler {params.assembler} ----minlen {params.len}  -R1 {input.read1} -R2 {input.read2} --outdir shovill_spades-{params.prefix} 2>&1> {log} 
        mv shovill_spades-{params.prefix} assemblies/shovill_spades
        """
##denove assembly with isolate spades
rule spades:
    input:
        read1 = "art_reads/{sample}_1.fq",
        read2 = "art_reads/{sample}_2.fq"
    output:
        output = "assemblies/spades/spades-{sample}/contigs.fasta"
    params:
        prefix = "{sample}",
        basic = config["spades"]["basic_options"],
        length = config["spades"]["min_contig_len"]
    conda:
        "envs/spades.yaml"
    log:
        "logs/spades/{sample}.log"
    threads: 8
    shell:
        """        
        spades.py -t {threads} {params.basic} -1 {input.read1} -2 {input.read2} -o spades-{params.prefix} 2>&1> {log}
        mv spades-{params.prefix} assemblies/spades            
        """
##prokka
rule prokka:
    input:
        reference = "genomes/{sample}.fasta"
    output:
        gff = "prokka/prokka-{sample}/prokka-{sample}.gff",
        tsv = "prokka/prokka-{sample}/prokka-{sample}.tsv",
    params:
        genus = config["prokka"]["genus"],
        prefix = "{sample}"
    conda:
        "envs/prokka.yaml"
    log:
        "logs/prokka/{sample}.log"
    threads: 2
    shell:
        """
        prokka --prefix prokka-{params.prefix} --genus {params.genus} --usegenus --fast --noanno --quiet --force --cpus {threads} {input.reference} 2>&1> {log}
        mv prokka-{params.prefix} prokka
        """
##quast for skesa
rule quast_skesa:
    input:
        contigs = "assemblies/skesa/{sample}.fasta",
        reference = "genomes/{sample}.fasta",
        ggf = "prokka/prokka-{sample}/prokka-{sample}.gff"
    output:
        tsv = "quast/quast_skesa-{sample}/report.tsv"
    params:
        prefix = "{sample}",
        min_len = config["quast"]["min_contig_len"]
    log:
        "logs/quast_skesa/{sample}.log"
    threads: 2
    conda:
        "envs/quast.yaml"
    shell:
        """        
        quast -t {threads} -m {params.min_len} {input.contigs} -r {input.reference} -g {input.ggf} -o quast_skesa-{params.prefix} 2>&1> {log}
        mv quast_skesa-{params.prefix} quast
        """
##quast for shovill
rule quast_shovill_velvet:
    input:
        contigs = "assemblies/shovill_velvet/shovill_velvet-{sample}/velvet.fasta",
        reference = "genomes/{sample}.fasta",
        ggf = "prokka/prokka-{sample}/prokka-{sample}.gff"
    output:
        tsv = "quast/quast_shovill_velvet-{sample}/report.tsv"
    params:
        prefix = "{sample}",
        min_len = config["quast"]["min_contig_len"]
    log:
        "logs/quast_shovill_velvet/{sample}.log"
    threads: 2
    conda:
        "envs/quast.yaml"
    shell:
        """      
        quast -t {threads} -m {params.min_len} {input.contigs} -r {input.reference} -g {input.ggf} -o quast_shovill_velvet-{params.prefix} 2>&1> {log}
        mv quast_shovill_velvet-{params.prefix} quast
        """

##quast for shovill_spades
rule quast_shovill_spades:
    input:
        contigs = "assemblies/shovill_spades/shovill_spades-{sample}/spades.fasta",
        reference = "genomes/{sample}.fasta",
        ggf = "prokka/prokka-{sample}/prokka-{sample}.gff"
    output:
        tsv = "quast/quast_shovill_spades-{sample}/report.tsv"
    params:
        prefix = "{sample}",
        min_len = config["quast"]["min_contig_len"]
    log:
        "logs/quast_shovill_spades/{sample}.log"
    threads: 2
    conda:
        "envs/quast.yaml"
    shell:
        """      
        quast -t {threads} -m {params.min_len} {input.contigs} -r {input.reference} -g {input.ggf} -o quast_shovill_spades-{params.prefix} 2>&1> {log}
        mv quast_shovill_spades-{params.prefix} quast
        """

##quast for spades
rule quast_spades:
    input:
        contigs = "assemblies/spades/spades-{sample}/contigs.fasta",
        reference = "genomes/{sample}.fasta",
        ggf = "prokka/prokka-{sample}/prokka-{sample}.gff"
    output:
        tsv = "quast/quast_spades-{sample}/report.tsv"
    params:
        prefix = "{sample}",
        min_len = config["quast"]["min_contig_len"]
    log:
        "logs/quast_spades/{sample}.log"
    threads: 2
    conda:
        "envs/quast.yaml"
    shell:
        """
        quast -t {threads} -m {params.min_len} {input.contigs} -r {input.reference} -g {input.ggf} -o quast_spades-{params.prefix} 2>&1> {log}
        mv quast_spades-{params.prefix} quast
        """

rule quast_report:
    input:
        expand("quast/quast_spades-{sample}/report.tsv", sample = IDS),
        expand("quast/quast_skesa-{sample}/report.tsv", sample= IDS),
        expand("quast/quast_shovill_velvet-{sample}/report.tsv", sample = IDS),
        expand("quast/quast_shovill_spades-{sample}/report.tsv", sample = IDS),
        expand("prokka/prokka-{sample}/prokka-{sample}.tsv", sample = IDS)
    output:
        spades = "quast-report/quast-report_spades.tsv",
        skesa = "quast-report/quast-report_skesa.tsv",
        shovill_v = "quast-report/quast-report_shovill_velvet.tsv",
        shovill_s = "quast-report/quast-report_shovill_spades.tsv",
        all = "quast-report/quast-report.tsv"
    params:
        quantity = config["quast_report"]["quantity"],
        name = config["quast_report"]["name"]
    conda:
        "envs/quast_report.yaml"
    log:
       "logs/quast_report/log"
    shell:
        """
        python Modules/quast-report.py -q {params.quantity} -o quast-report/{params.name} 2>&1> {log}
        """
rule boxplot:
    input:
        df = "quast-report/quast-report.tsv"
    output:
        NGA50_pdf = "plots/NGA50.pdf",
        contigs_pdf = "plots/contigs.pdf",
        missassemblies_pdf = "plots/misassemblies.pdf",
        fraction_pdf = "plots/genome_fraction.pdf",
        c_features_pdf = "plots/complete_features.pdf",
        t_features_pdf = "plots/total_features.pdf"
    conda:
        "envs/R.yaml"
    log:
       "logs/boxplot.log"
    shell:
         """
         Rscript Modules/boxplot.R -f {input.df} 2>&1> {log}
         mv *.pdf plots/
         """

rule SKA_kmer_reads:
    input:
        read1 = "art_reads/{sample}_1.fq",
        read2 = "art_reads/{sample}_2.fq"
    output:
        ska = "ska_fastq_out/{sample}.skf"
    params:
        strain = "ska_fastq_out/{sample}"

    threads: 8
    conda:
         "envs/ska.yaml"
    log:
       "logs/ska/kmer_fastq/{sample}.log"
    shell:
        """
        ska fastq -o {params.strain} {input.read1} {input.read2} 2>&1>{log}
        """

rule SKA_kmer_shovill:
    input:
        fasta = "assemblies/shovill_velvet/shovill_velvet-{sample}/velvet.fasta"
    output:
        ska = "ska_shovill_out/{sample}.skf"
    params:
        strain = "ska_shovill_out/{sample}"
    threads: 8
    conda:
         "envs/ska.yaml"
    log:
       "logs/ska/kmer_shovill/{sample}.log"
    shell:
        """
        ska fasta -o {params.strain} {input.fasta} 2>&1>{log}
        """

rule SKA_kmer_spades:
    input:
        fasta = "assemblies/spades/spades-{sample}/contigs.fasta"
    output:
        ska = "ska_spades_out/{sample}.skf"
    params:
        strain = "ska_spades_out/{sample}"
    threads: 8
    conda:
         "envs/ska.yaml"
    log:
       "logs/ska/kmer_spades/{sample}.log"
    shell:
        """
        ska fasta -o {params.strain} {input.fasta} 2>&1>{log}
        """
rule SKA_kmer_skesa:
    input:
        fasta = "assemblies/skesa/{sample}.fasta"
    output:
        ska = "ska_skesa_out/{sample}.skf"
    params:
        strain = "ska_skesa_out/{sample}"
    threads: 8
    conda:
         "envs/ska.yaml"
    log:
       "logs/ska/kmer_skesa/{sample}.log"
    shell:
        """
        ska fasta -o {params.strain} {input.fasta} 2>&1>{log}
        """


rule SKA_align_reads:
    input:
        fastq = expand("ska_fastq_out/{sample}.skf", sample = IDS)
    output:
        aln = "ska_output/ska_fastq_alignment.aln"
    params:
        output= "ska_output/ska_fastq_alignment"
    threads: 8
    conda:
         "envs/ska.yaml"
    log:
       "logs/ska/align/fastq.log"
    shell:
        """
        ska align -o {params.output} -p 0 {input.fastq} 2>&1>{log}
        """

rule SKA_align_shovill:
    input:
        fastq = expand("ska_shovill_out/{sample}.skf", sample = IDS)
    output:
        aln = "ska_output/ska_shovill_alignment.aln"
    params:
        output= "ska_output/ska_shovill_alignment"
    threads: 8
    conda:
         "envs/ska.yaml"
    log:
       "logs/ska/align/shovill_fasta.log"
    shell:
        """
        ska align -o {params.output} -p 0 {input.fastq} 2>&1>{log}
        """
rule Ska_align_spades:
    input:
        fastq = expand("ska_spades_out/{sample}.skf", sample = IDS)
    output:
        aln = "ska_output/ska_spades_alignment.aln"
    params:
        output= "ska_output/ska_spades_alignment"
    threads: 8
    conda:
         "envs/ska.yaml"
    log:
       "logs/ska/align/spades_fasta.log"
    shell:
        """
        ska align -o {params.output} -p 0 {input.fastq} 2>&1>{log}
        """
rule Ska_align_skesa:
    input:
        fastq = expand("ska_skesa_out/{sample}.skf", sample = IDS)
    output:
        aln = "ska_output/ska_skesa_alignment.aln"
    params:
        output= "ska_output/ska_skesa_alignment"
    threads: 8
    conda:
         "envs/ska.yaml"
    log:
       "logs/ska/align/skesa_fasta.log"
    shell:
        """
        ska align -o {params.output} -p 0 {input.fastq} 2>&1>{log}
        """
rule prokka_spades:
    input:
        spades = "assemblies/spades/spades-{sample}/contigs.fasta"
    output:
        prokka =  "prokka/spades/spades-{sample}/spades-{sample}.gff"
    threads: 8
    params:
        genus = config["prokka"]["genus"],
        prefix = "{sample}"
    log:
        "logs/prokka/spades/{sample}.log"
    conda:
        "envs/prokka.yaml"
    shell:
        """
        prokka --prefix spades-{params.prefix} --centre X --compliant --genus {params.genus} --usegenus --quiet --force --cpus {threads} {input.spades} 2>&1> {log}
        mv spades-{params.prefix} prokka/spades
        """


rule prokka_skesa:
    input:
        skesa = "assemblies/skesa/{sample}.fasta"
    output:
        prokka =  "prokka/skesa/skesa-{sample}/skesa-{sample}.gff"
    threads: 8
    params:
        genus = config["prokka"]["genus"],
        prefix = "{sample}"
    log:
        "logs/prokka/skesa/{sample}.log"
    conda:
        "envs/prokka.yaml"
    shell:
        """
        prokka --prefix skesa-{params.prefix} --centre X --compliant --genus {params.genus} --usegenus --quiet --force --cpus {threads} {input.skesa} 2>&1> {log}
        mv skesa-{params.prefix} prokka/skesa
        """

rule prokka_shovill:
    input:
        shovill = "assemblies/shovill_velvet/shovill_velvet-{sample}/velvet.fasta"
    output:
        prokka =  "prokka/shovill/shovill-{sample}/shovill-{sample}.gff"
    threads: 8
    params:
        genus = config["prokka"]["genus"],
        prefix = "{sample}"
    log:
        "logs/prokka/shovill/{sample}.log"
    conda:
        "envs/prokka.yaml"
    shell:
        """
        prokka --prefix shovill-{params.prefix} --centre X --compliant --genus {params.genus} --usegenus --quiet --force --cpus {threads} {input.shovill} 2>&1> {log}
        mv shovill-{params.prefix} prokka/shovill
        """
rule roary_spades:
    input:
         prokka = expand("prokka/spades/spades-{sample}/spades-{sample}.gff", sample = IDS)
    output:
          aln = "roary/spades/core_gene_alignment.aln"
    threads: 8
    log:
        "logs/roary/spades.log"
    conda:
        "envs/roary.yaml"
    shell:
        """
        roary -e --mafft {threads} {input.prokka} -f roary/spades 2>&1> {log}
        """
rule roary_skesa:
    input:
         prokka = expand("prokka/skesa/skesa-{sample}/skesa-{sample}.gff", sample = IDS)
    output:
          aln = "roary/skesa/core_gene_alignment.aln"
    threads: 8
    log:
        "logs/roary/skesa.log"
    conda:
        "envs/roary.yaml"
    shell:
        """
        roary -e --mafft {threads} {input.prokka} -f roary/skesa 2>&1> {log}
        """
rule roary_shovill:
    input:
         prokka = expand("prokka/shovill/shovill-{sample}/shovill-{sample}.gff", sample = IDS)
    output:
          aln = "roary/shovill/core_gene_alignment.aln"
    threads: 8
    log:
        "logs/roary/shovill.log"
    conda:
        "envs/roary.yaml"
    shell:
        """
        roary -e --mafft {threads} {input.prokka} -f roary/shovill 2>&1> {log}
        """
rule pirate_spades:
    input:
         prokka = expand("prokka/shovill/shovill-{sample}/shovill-{sample}.gff", sample = IDS)
    output:
          aln = "pirate/spades/core_alignment/pangenome_alignment.fasta"
    threads: 8
    params:
        dir = "prokka/spades-prokka"
    log:
        "logs/roary/shovill.log"
    conda:
        "envs/roary.yaml"
    shell:
        """
        mkdir {params.dir}
        cp {input.prokka} {params.dir}
        pirate -i {params.dir} -o pirate/spades -a 2>&1> {log}
        """

rule pirate_skesa:
    input:
        prokka = expand("prokka/skesa/skesa-{sample}/skesa-{sample}.gff", sample = IDS)
    output:
        aln = "pirate/skesa/core_alignment/pangenome_alignment.fasta"
    threads: 8
    params:
        dir = "prokka/skesa-prokka"
    log:
        "logs/roary/skesa.log"
    conda:
        "envs/pirate.yaml"
    shell:
        """
        mkdir {params.dir}
        cp {input.prokka} {params.dir}
        pirate -i {params.dir} -o pirate/skesa -a 2>&1> {log}
        """

rule pirate_shovill:
    input:
        prokka = expand("prokka/shovill/shovill-{sample}/shovill-{sample}.gff", sample = IDS)
    output:
        aln = "pirate/shovill/core_alignment/pangenome_alignment.fasta"
    threads: 8
    params:
        dir = "prokka/shovill-prokka"
    log:
        "logs/roary/shovill.log"
    conda:
        "envs/pirate.yaml"
    shell:
        """
        mkdir {params.dir}
        cp {input.prokka} {params.dir}
        pirate -i {params.dir} -o pirate/shovill -a 2>&1> {log}
        """
rule snippy:
    input:
         read1 = "art_reads/{sample}_1.fq",
         read2 = "art_reads/{sample}_2.fq",
         ecolik12 = config["snippy"]["ecolik12"]
    output:
         aln = directory("snippy/{sample}")
    params:
        general = config["snippy"]["general"]
    threads: 8
    log:
         "logs/snippy/snippy-{sample}.log"
    conda:
         "envs/snippy.yaml"
    shell:
         """
         snippy {params.general} --cpus {threads} --outdir {output.aln} --ref {input.ecolik12} --pe1 {input.read1} --pe2 {input.read2} 2>{log} 
         """

rule snippy_core:
    input:
        data = expand("snippy/{sample}", sample=IDS),
        ecolik12 = config["snippy"]["ecolik12"]
    output:
        full = "snippy-core_out/core.full.aln",
        snps = "snippy-core_out/core.aln"
    params:
         general = config["snippy"]["general"],
         outdir = "snippy_core_out"
    threads: 8
    log:
        "logs/snippy/snippy-core.log"
    conda:
        "envs/snippy.yaml"
    shell:
        """
        mkdir -p {params.outdir}
		snippy-core {params.general} --cpus {threads} --ref {input.ecolik12} {input.data} 2>&1>{log}
		mv core.aln core.full.aln {params.outdir}
        """
rule ksnp_make_skesa:
    input:
        contigs = expand("assemblies/skesa/{sample}.fasta", sample=IDS)
    output:
        list = "list_skesa_kSNP"
    conda:
        "envs/ksnp.yaml"
    params:
        "assemblies/skesa"
    shell:
        """
        MakeKSNP3infile {params} {output.list} A
        """
rule ksnp_dir_spades:
    input:
        contigs = "assemblies/spades/spades-{sample}/contigs.fasta"
    output:
        dir = "assemblies/spades_dir/{sample}.fasta"
    params:
        "{sample}"
    shell:
        """
        cp {input.contigs} {output.dir}
        """
rule ksnp_dir_shovill:
    input:
        contigs = "assemblies/shovill_velvet/shovill_velvet-{sample}/velvet.fasta"
    output:
        dir = "assemblies/shovill_dir/{sample}.fasta"
    params:
        "{sample}"
    shell:
        """
        cp {input.contigs} {output.dir}
        """

rule ksnp_make_spades:
    input:
        contigs = expand("assemblies/spades_dir/{sample}.fasta", sample=IDS)
    output:
        list = "list_spades_kSNP"
    conda:
        "envs/ksnp.yaml"
    params:
        "assemblies/spades_dir"
    shell:
        """
        MakeKSNP3infile {params} {output.list} A
        """
rule ksnp_make_shovill:
    input:
        contigs = expand("assemblies/shovill_dir/{sample}.fasta", sample=IDS)
    output:
        list = "list_shovill_kSNP"
    conda:
        "envs/ksnp.yaml"
    params:
        "assemblies/shovill_dir"
    shell:
        """
        MakeKSNP3infile {params} {output.list} A
        """
rule ksnp_skesa:
    input:
        list = "list_skesa_kSNP"
    output:
        kSNP3 = "kSNP_skesa/core_SNPs_matrix"
    conda:
        "envs/ksnp.yaml"
    threads: 8
    params:
        kmersize = config["ksnp"]["kmersize"]
    log:
        "logs/ksnp_skesa.log"
    shell:
        """
        kSNP3 -in {input.list} -outdir {output.kSNP3} -k {params.kmersize} -core -CPU {threads} 2>&1>{log} 
        """
rule ksnp_spades:
    input:
        list = "list_spades_kSNP"
    output:
        kSNP3 = "kSNP_spades/core_SNPs_matrix"
    conda:
        "envs/ksnp.yaml"
    threads: 8
    params:
        kmersize = config["ksnp"]["kmersize"]
    log:
        "logs/ksnp_spades.log"
    shell:
        """
        kSNP3 -in {input.list} -outdir {output.kSNP3} -k {params.kmersize} -core -CPU {threads} 2>&1>{log} 
        """
rule ksnp_shovill:
    input:
        list = "list_shovill_kSNP"
    output:
        kSNP3 = "kSNP_shovill/core_SNPs_matrix"
    conda:
        "envs/ksnp.yaml"
    threads: 8
    params:
        kmersize = config["ksnp"]["kmersize"]
    log:
        "logs/ksnp_shovill.log"
    shell:
        """
        kSNP3 -in {input.list} -outdir {output.kSNP3} -k {params.kmersize} -core -CPU {threads} 2>&1>{log} 
        """



